Satellite module to manage activation keys

liquidat opened the pull request:
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

satellite_activation_keys (in the cloud subdirectory)
##### ANSIBLE VERSION

```
ansible 2.0.2.0
```
##### SUMMARY

This is a module to manage Satellite, specifically the so called _activation keys_. The technical basis is to manage everything via REST calls, the Python module needed for the REST API is the requests module.

If the module is accepted, the plan is to use it as a blueprint for other Satellite modules since it already implements:
- structure to generally query Satellite via REST API
- sub functions to lookup core information in Satellite like organizations, environments, content views, etc.

From the commit changelog:

DONE:
- manages Red Hat Satellite and thus Katello REST API calls
- uses Python requests library
- all major parameters for create and update are covered
- check_mode supported
- catches errors when parameters cannot be found
- no pep8 errors

TODO:
- Return is sometimes JSON code, sometimes string, could be improved

gregdek commented:
Thanks @liquidat for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

liquidat commented:
I've fixed all build problems, namely:
- replaced requests library with open_url (yes, that was a bit hard)
- fixed exceptions so the Python 3 tests are ok
- fixed Return description, should be string only now all the time

gregdek commented:
@liquidat A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

liquidat commented:
@gregdek Please help me out here, I have no idea what is still missing and how to let you know via PR that everything is fine!

tima commented:
ready_for_review

(@liquidat: I think that will get the attention of the ansibull bot)

gregdek commented:
Hmm, got wedged for some reason. Kicking.

gregdek commented:
Thanks @liquidat for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

tima commented:
This PR is related to (but not dependant upon) #2300.

liquidat commented:
@tima I've fixed the naming as you suggested. However, I am not sure why the build now fails, looks to me as a problem of the build system...?

woot4moo commented on cloud/satellite/satellite_activation_keys.py:
```diff
@@ -0,0 +1,365 @@
+    """Looks up content view ID for a given content view name.
+    """
+
+    url_action = 'https://' + module.params.get('satellite') + '/katello/api/v2/content_views/'
```
it seems this could safely be moved out to a constant as it stands this is a magic string this applies to the next several lines as well

liquidat replied:
@woot4moo You mean the part with `module.params.get('host')`? I thought about moving that to a constant and handing it over as a parameter in each function call.
However, at the same time the module itself is already handed over at each function, thus the value is already in the function. I thought I have less duplication this way - but could change it if you prefer.

woot4moo replied:
I have learned to accept that some duplicate code is OK if the trade-off is it makes the code easier to remove.  What I see happening here is we are hard coding some values that make our code more open to the possibility of duplication without the benefits of easier to delete.  I would move it into a constant for this purpose

liquidat replied:
Ah, ok, thanks for the clarification. I will change the code as together with the kwargs idea from below (if I got that right).

liquidat replied:
Done.

woot4moo commented on cloud/satellite/satellite_activation_keys.py:
```diff
@@ -0,0 +1,365 @@
+    resp = open_url(url_action,method="GET",headers=headers,url_username=module.params.get('user'),url_password=module.params.get('pwd'),force_basic_auth=True,data=json.dumps(payload),validate_certs=module.params.get('validate_certs'))
+    resp_json = json.loads(resp.read())
+
+    if resp_json["total"] == 0:
```
this error handling code is duplicate, perhaps move this into a small helper function

liquidat replied:
The problem is that there are small, but important differences between error handling in the functions. It would require at least two small helper functions, but I will look into it.

woot4moo commented on cloud/satellite/satellite_activation_keys.py:
```diff
@@ -0,0 +1,365 @@
+    organization      = module.params.get('organization')
+    key_action  = module.params.get('key_action')
+    key_name  = module.params.get('key_name')
+    validate_certs = module.params.get('validate_certs')
```
**kwargs fits nicely here

liquidat replied:
@woot4moo How exactly would you bring it in here? Do you mean to not hand over the module object, but to hand over the variables (user, pwd, etc.) for **kwargs?

liquidat replied:
@woot4moo I am still not sure where to introduce kwargs in the best way: the manage_key function still needs to have the entire module, right? Or would you already hand over the parameters to this function via kwargs?

woot4moo commented on cloud/satellite/satellite_activation_keys.py:
```diff
@@ -0,0 +1,365 @@
+            module.fail_json(msg="Key name %s not found." % key_name)
+        url_action_put = url_action + str(key["id"])
+        payload = {"organization_id":orga_id}
+        if ('key_description' in locals()) and (key_description != key["description"]):
```
what does locals return?  Ideally a dict, but if this is a list the complexity on this lookup is O(n)

liquidat replied:
@woot4moo It returns a dictionary representing the current local symbol table. My problem here is that many module parameters are optional and I need to find out if the parameter was set - or not.

gundalow commented:
Hi,

Looks like this branch needs updating to fix the Travis issue, please see this guide on how to do that cleanly.

http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

needs_revision

gregdek commented:
Thanks @liquidat for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@liquidat A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@liquidat Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@liquidat A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

liquidat commented:
I have addressed all issues mentioned above. In some cases, like the idea to usage of kwargs, I haven't changed anything since the reported did not reply to my request for further details.

ready_for_review

gregdek commented:
Thanks @liquidat for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]
